### PR TITLE
Add new recommendation algorithm FMA

### PR DESCRIPTION
**Pull request by jilljenn**
FMA = Factorization Machine = FullMetal Alchemist.

@RaitoBezarius va hurler, mais c'est pour le bien de tous.

### CODE REVIEW

File: mangaki/mangaki/algo/fma.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        X_fm = coo_matrix(([1] * (2 * nb_samples), (rows, cols)),
+                          shape=(nb_samples, self.nb_users + self.nb_works)
+                          ).tocsr()
+        return X_fm
+
+    def fit(self, X, y):
+        X_fm = self.prepare_fm(X)
+        self.chrono.save('prepare data in sparse FM format')
+
+        os.environ['LIBFM_PATH'] = 'XXX'  # If applicable
```
Comment (RaitoBezarius) [inline mangaki/mangaki/algo/fma.py]: :-[

File: mangaki/mangaki/algo/fma.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        fm = pywFM.FM(task='regression', num_iter=self.nb_iterations, k2=self.rank, rlog=False)  # MCMC method
+        model = fm.run(X_fm, y, X_fm, y)
+        self.chrono.save('train FM' + str(model))
+
+        NB_AGENTS = self.nb_users + self.nb_works
```
Comment (RaitoBezarius) [inline mangaki/mangaki/algo/fma.py]: Could we have the lowercase version?

File: mangaki/mangaki/algo/fma.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        current = len(model.weights)
+
+        if model.global_bias is None:  # Train failed (for example, libfm does not exist)
+            self.mu = 0
```
Comment (RaitoBezarius) [inline mangaki/mangaki/algo/fma.py]: How do I know that my FMA model is, in fact, a baseline stupid model when you do this?
Could you just raise an exception? Wouldn't it be better to inform the user of unexpected errors?

File: mangaki/mangaki/algo/fma.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    def load(self, filename):
+        backup = super().load(filename)
+        self.mu = backup.mu
+        self.W = backup.W
+        self.V = backup.V
+        self.V2 = backup.V2
+
+    @property
+    def is_serializable(self):
+        return False  # Not yet
```
Comment (RaitoBezarius) [inline mangaki/mangaki/algo/fma.py]: Nit: if it is false, it's obvious this is not yet serializable :p.

File: mangaki/mangaki/algo/fma.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                          shape=(nb_samples, self.nb_users + self.nb_works)
+                          ).tocsr()
+        return X_fm
+
+    def fit(self, X, y):
+        X_fm = self.prepare_fm(X)
+        self.chrono.save('prepare data in sparse FM format')
+
+        os.environ['LIBFM_PATH'] = 'XXX'  # If applicable
+        fm = pywFM.FM(task='regression', num_iter=self.nb_iterations, k2=self.rank, rlog=False)  # MCMC method
```
Comment (RaitoBezarius) [inline mangaki/mangaki/algo/fma.py]: Could we parametrize `rlog`? And elaborate on this parameter?

File: mangaki/mangaki/algo/fma.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        model = fm.run(X_fm, y, X_fm, y)
+        self.chrono.save('train FM' + str(model))
```
Comment (RaitoBezarius) [inline mangaki/mangaki/algo/fma.py]: Nit: missing space after `FM`.

File: mangaki/mangaki/algo/fma.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+        if model.global_bias is None:  # Train failed (for example, libfm does not exist)
+            self.mu = 0
+            self.W = np.random.random(NB_AGENTS)
+            self.V = np.random.random((NB_AGENTS, self.rank))
+        else:
+            self.mu = model.global_bias
+            self.W = np.pad(np.array(model.weights), (0, NB_AGENTS - current), mode='constant')  # Just in case X_fm had too many zero columns on the right
+            self.V = np.pad(model.pairwise_interactions, [(0, NB_AGENTS - current), (0, 0)], mode='constant')
+        self.V2 = np.power(self.V, 2)
```
Comment (RaitoBezarius) [inline mangaki/mangaki/algo/fma.py]: Curious: Is this not the same as `self.V ** 2`? Does that version run faster due to NumPy explicit usage?

File: mangaki/mangaki/algo/fma.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    @property
+    def is_serializable(self):
+        return False  # Not yet
+
+    def prepare_fm(self, X):
+        nb_samples = len(X)
+        user_ids = X[:, 0]
+        work_ids = X[:, 1]
+        # For the k-th user_id-work_id pair, we need (k, user_id) and (k, N + work_id), so two copies of range(nb_samples)
+        rows = list(range(nb_samples)) * 2
```
Comment (RaitoBezarius) [inline mangaki/mangaki/algo/fma.py]: Wonder if we could do it faster using NumPy built-ins for `range` and so on?

File: mangaki/mangaki/algo/fma.py (hunk 1 above)
Comment (jilljenn) [reply mangaki/mangaki/algo/fma.py]: Sois pas trop inquiet, c'est un WIP. (Même si c'est pas inscrit comme tel.)

Vais peut-être mettre ce qu'il faut dans `settings.py`, mais il faut qu'on en discute plus bas.

File: mangaki/mangaki/algo/fma.py (hunk 2 above)
Comment (jilljenn) [reply mangaki/mangaki/algo/fma.py]: Done.

File: mangaki/mangaki/algo/fma.py (hunk 3 above)
Comment (jilljenn) [reply mangaki/mangaki/algo/fma.py]: I do it to pass the tests. More below.

File: mangaki/mangaki/algo/fma.py (hunk 4 above)
Comment (jilljenn) [reply mangaki/mangaki/algo/fma.py]: Actually, now it is.

Comment (jilljenn) [reply mangaki/mangaki/algo/fma.py]: It was from CFM.

File: mangaki/mangaki/algo/fma.py (hunk 5 above)
Comment (jilljenn) [reply mangaki/mangaki/algo/fma.py]: I use it for my research now. I will add a comment.

File: mangaki/mangaki/algo/fma.py (hunk 6 above)
Comment (jilljenn) [reply mangaki/mangaki/algo/fma.py]: Actually, `str(model)` was debug. I removed it.

File: mangaki/mangaki/algo/fma.py (hunk 7 above)
Comment (jilljenn) [reply mangaki/mangaki/algo/fma.py]: I prefer to precompute it, then every predict operation becomes easy.

File: mangaki/mangaki/algo/fma.py (hunk 8 above)
Comment (jilljenn) [reply mangaki/mangaki/algo/fma.py]: Good question. I will provide a better implementation, actually, anyway.

**Pull request merged**